src/proof/alethe/alethe_post_processor.cpp

```diff
@@ -1539,30 +1539,79 @@ bool AletheProofPostprocessCallback::update(Node res,
     //
-    // In case the sigma is the identity this step is merely converted to
+    // When sigma is the substitution {y1->z1, ..., yn->zn, ..., yn+k->zn+k}, we
+    // are in the case where G has quantifiers whose variables are yn+1 ... yn+k
+    // and they will be renamed to zn+1 ... zn+k. The generated Alethe proof is
+    //
+    //  --------------------------------------- refl
+    //  (cl (= G G*{y1 -> z1, ..., yn -> zn}))
+    // --------------------------------- bind, z1 ... zn (= y1 z1) ... (= yn zn)
+    //  (cl (= F (forall ((z1 A1) ... (zn An)) G*{y1 -> z1, ..., yn -> zn})))
+    //
+    // i.e., we drop the suffix of the substitution beyond the variables of the
+    // outermost quantifier. This is valid in Alethe because the validity of
+    // "refl", under a rule that introduces a context, such as "bind", is itself
+    // tested modulo alpha-equivalence. An alternative would be to use the rest
+    // of the substitution to do new "bind" steps for the innermost quantifiers.
+    //
+    // If sigma contains more than one variable with the same name but with
+    // different types (which makes them different for cvc5 but not in the
+    // substitution induced by Alethe's context), we introduce an intermediate
+    // alpha equivalence step with fresh variables. For example if F := (forall
+    // ((y1 A1) (y2 A2)) G) and sigma is the substitution {y1 -> z1, y2 -> y1},
+    // where the "y1" in the right hand side has another type "T" other than
+    // "A2", then the resulting steps are
+    //
+    //  --------------------------------------- refl
+    //  (cl (= G G*{y1 -> @v1, y2 -> @v2})
+    // --------------------------------- bind, @v1 @v2 (= y1 @v1) (= y1 @v2)
+    //  (cl (= F (forall ((@v1 A1) (@v2 A2)) G*{y1 -> @v1, y2 -> @v2})
+    //
+    //  --------------------------------------------------- refl
+    //  (cl (= G*{y1->@v1, y2->@v2} (G*{y1->@v1, y2->@v2})*{@v1->z1, @v1->y1})
+    // --------------------------------------- bind, z1 y1 (= @v1 z1) (= @v2 z2)
+    //  (cl (=
+    //    (forall ((@v1 A1) (@v2 A2)) G*{y1 -> @v1, y2 -> @v2})
+    //    (forall ((z1 A1) (y1 A2)) (G*{y1->@v1, y2->@v2})*{@v1->z1, @v1->y1})))
+    //
+    // and finally a transitivity step to conclude an equality between F and
+    // (forall ((z1 A1) (y1 A2)) (G*{y1->@v1, y2->@v2})*{@v1->z1, @v1->y1}).
+    //
+    // If none of the conditions above applies, the resulting Alethe steps are
     //
     //  ------------------ refl
-    //  (cl (= F F))
+    //  (cl (= G G*sigma))
+    // -------------------- bind, z1 ... zn (= y1 z1) ... (= yn zn)
+    //  (cl (= F (forall ((z1 A1) ... (zn An)) G*sigma)))
     case ProofRule::ALPHA_EQUIV:
     {
-      std::vector<Node> varEqs;
       // If y1 ... yn are mapped to y1 ... yn it suffices to use a refl step
       bool allSame = true;
+      std::vector<std::string> lhsNames;
+      bool hasClash = false;
       for (size_t i = 0, size = res[0][0].getNumChildren(); i < size; ++i)
```

Comment:
This would not work for

```
res[0][0] = ((?v0 A) (?v1 B)) 
res[1][0] = ((?v1 A) (?v2 B))
```
Or am I overlooking something?

Reply:
You're right, it was a bug. Fixed and added a regression.